### ISSUE TITLE
Domino: Using "Default Value" from Formula instead "Value" 

### ISSUE THREAD

**Issue by piotrzarzycki21** [Moonshine-IDE/Moonshine-IDE#1069]
Some bigger changes that are still pending before we can start on integration with the agents:
- View Builder (#1057) - I need to revisit this soon
- Default Value.  Currently this is populated from the "Value" property, but this is not correct for Domino Visual Editor.  The actual default value is based on a Domino formula (Field Info > Client Objects > defaultvalue > Formula).  Don't worry about parsing this formula for now - it will be difficult to handle cases beyond no value and a string or number literal.   The text default value will cause problems for at least the type=number fields as we move to the agents.  

![image](https://user-images.githubusercontent.com/24532731/180881206-da5384d2-f7e5-460e-9807-f3f6d7cdc1da.png)

_Originally posted by @JoelProminic in https://github.com/Moonshine-IDE/Moonshine-IDE/issues/1040#issuecomment-1194685251_

**Comment by piotrzarzycki21** [comment]
That "Formula" seems to be not present in our intermediate XML: 

`<Field height="30" hide="" hidewhen="" htmlClass="" htmlId="" htmlOther="" htmlStyle="" htmlTitle="" id="" kind="editable" listdisplayseparator="" listinputseparators="" maxlength="" name="HostName" required="false" type="text" value="HostName" width="100"/>`

I checked also corresponding Domino code and it is also not present:

`<Field height="30" hide="" hidewhen="" htmlClass="" htmlId="" htmlOther="" htmlStyle="" htmlTitle="" id="" kind="editable" listdisplayseparator="" listinputseparators="" maxlength="" name="HostName" required="false" type="text" value="HostName" width="100"/>`

@JoelProminic Is it an issue or I'm missing something ?

**Comment by JoelProminic** [comment]
No formula indicates that the default value should be "" or unset (for numbers/DateTime/keyword).

Looking through the PerformanceStats database, I actually don't see an example of a default value.    I'll need to provide you another example database for this.

**Comment by JoelProminic** [comment]
While discussing a [separate issue](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/1064#issuecomment-1201261592), @feather812002 reported that the Value property is currently only used by Domino Visual Editor to set the hint text on the field (if this is not set, then it will show "Input Text" when the form is reopened):
![image](https://user-images.githubusercontent.com/24532731/182483104-7886ff3b-0c4a-487d-a342-b95e99c64314.png)

However: 
- this hint text is controlled by `Advanced > Field Hint` for the actual Domino output.
- The default value for the Domino field is controlled by `Field Info > Client Objects > defaultvalue > Formula`

Should we just get rid of (or hide) the Value field in Domino Visaul editor?  Is there any other purpose here that we are forgetting?

**Comment by piotrzarzycki21** [comment]
@JoelProminic I have made changes where I'm using following rule to display stuff in Domino Text input:
`fieldValue = this.formula ? this.formula : this.defaultvalue;`

Is it correct ? I think formula is not parsed correctly and I don't get any value, so this need to be fixed at some point. 

Because properties in VO has now empty strings as an initial it end up with displaying empty rows in Grids.

**Comment by piotrzarzycki21** [comment]
> Is it correct ? I think formula is not parsed correctly and I don't get any value, so this need to be fixed at some point.

There is also possibility that something is wrong on my sight - I'm investigating that.

**Comment by JoelProminic** [comment]
Sorry @piotrzarzycki21, I haven't been seeing your updates, and it took me longer than intended to revisit this.

With the current strategry, the current "Value" property (`this.defaultValue`?) should not be used for Domino - we have other properties to handle this logic.

Instead, we want to use `Field Info > Client Objects > defaultvalue > Formula` (`this.formula`?), but the value can't necessarily be used directly in Royale, since it is a "formula".  For example:

Type | Multivalue | Example Literal
-----|------------|----------------
Text/Richtext | false | `"foo"`
Text | true | `"foo":"bar"`
Number | false | `123`
Number | true | `123:456`

DateTimes have literals as well, but I have never used these in Designer.  Instead I use formulas like `@Now`, `@Today` or `@Yesterday` to generate a useful value.

The formulas are code, so they can be arbitrarily complex:
```
value := @DbLookup("":"";"":""; "TestDefaultValue"; "MyDefault"; "TextLiteral");
@If (@IsError(value);
     "";
     value)
```

I don't expect the conversion code to be able to parse an arbitrary formula, so any formula that can't be parsed should be included in the code as a comment instead, for manual conversion.  **For now, you can just treat all of the formula values as "unparsable", and we'll move this logic to a separate issue.**

I tried to create a test database for this, but unfortunately it is not opening properly in Moonshine.  I'll need to investigate this more, but hopefully I clarified the issue.

**Comment by piotrzarzycki21** [comment]
I get to 

> Sorry @piotrzarzycki21, I haven't been seeing your updates, and it took me longer than intended to revisit this.
> 
> With the current strategry, the current "Value" property (`this.defaultValue`?) should not be used for Domino - we have other properties to handle this logic.
> 
> Instead, we want to use `Field Info > Client Objects > defaultvalue > Formula` (`this.formula`?), but the value can't necessarily be used directly in Royale, since it is a "formula". For example:
> 
> Type	Multivalue	Example Literal
> Text/Richtext	false	`"foo"`
> Text	true	`"foo":"bar"`
> Number	false	`123`
> Number	true	`123:456`
> DateTimes have literals as well, but I have never used these in Designer. Instead I use formulas like `@Now`, `@Today` or `@Yesterday` to generate a useful value.
> 
> The formulas are code, so they can be arbitrarily complex:
> 
> ```
> value := @DbLookup("":"";"":""; "TestDefaultValue"; "MyDefault"; "TextLiteral");
> @If (@IsError(value);
>      "";
>      value)
> ```
> 
> I don't expect the conversion code to be able to parse an arbitrary formula, so any formula that can't be parsed should be included in the code as a comment instead, for manual conversion. **For now, you can just treat all of the formula values as "unparsable", and we'll move this logic to a separate issue.**
> 
> I tried to create a test database for this, but unfortunately it is not opening properly in Moonshine. I'll need to investigate this more, but hopefully I clarified the issue.

This clarifies all of that. However I discovered that we are parsing what we have in formula field to defaultvalue property. If you don't have anything in formula -> defaultvalue is empty string, if you have something defaultvalue has some value. 
Basically property formula is not used at all. 

<img width="277" alt="Screenshot 2022-08-09 at 12 07 07" src="https://user-images.githubusercontent.com/24554795/183622943-7165973d-5938-4bfe-af6e-f10d36e92cb3.png">
<img width="264" alt="Screenshot 2022-08-09 at 12 04 23" src="https://user-images.githubusercontent.com/24554795/183622732-bfbae02d-4e1f-40d5-beeb-cfe069709129.png">


Depends what we are choosing in drop down list that property is being used - it took me quite a bit of time to figure out this. I think I should check each property whether it is empty - if it's not I should probably use it in my VO.

```
package vo
{
   [Bindable] 
   public class TestDominoVeProjectVO
   {
       public var test:String = "super super";

       public function copy():Object
       {
           var o:Object = new TestDominoVeProjectVO();
     			o.test = this.test;
           return o;
		}
   } 
}
```

**Comment by piotrzarzycki21** [comment]
Actually all of these options are being saved, so if some value is as part Input Translation and Default value both of the are going to be part of dxl. Right now I'm using default value in my exported project.

```
<field listdisplayseparator="newline" listinputseparators="comma semicolon newline" useappletinbrowser="false" allowtabout="false" defaultfocus="false" storelocally="false" alignPardef="left" inputtranslation="super super" value="eee" type="text" kind="editable" computeaftervalidation="false" allowmultivalues="false" width="100pt" height="30pt" bgcolor="#ffffff" name="test" sign="false" seal="false" protected="false">
            <code event="defaultvalue">
              <formula>super superddddff</formula>
            </code>
            <code event="inputtranslation">
              <formula>super super</formula>
            </code>
          </field>
```

**Comment by piotrzarzycki21** [comment]
After my changes right now Bing's project is failing, cause value of some fields have some special characters and Royale compiler complains that they are not allowed. 

**HostConfigSnapshotVO:**

```
       public var ZonesRunningNotAutoboot:String = "@Transform(ComputedRunningZones; "x";
           @If(@IsMember(x; ComputedZoneAutobootTrue); @Nothing; x))";
       public var ZonesNotRunningAutoboot:String = "@Transform(ComputedNonRunningZones; "x";
           @If(@IsMember(x; ComputedZoneAutobootTrue); x; @Nothing))";
```

**Error:**

```
HostConfigSnapshotVO.as(47): col: 90 Syntax error: A string literal must be terminated before the line break.
:        public var ZonesRunningNotAutoboot:String = "@Transform(ComputedRunningZones; "x";
:                                                                                          ^
```

**Comment by JoelProminic** [comment]
@piotrzarzycki21, to clarify, are you seeing errors for the formulas on the Domino Visual Editor side?  I see that the code snippet is invalid for Royale because of the unescaped double quotes.

To clarify, I don't want to put the formula as the default value for fields, since this will fill the form with undesired text that will need to be removed/replaced by the end user.  

Instead, the formula should be added as a comment, and the default value should be "".  We can make improvements later to parse simple formulas (i.e literals).

So in this case, the code should look something like:

```
       public var ZonesRunningNotAutoboot:String = "";  /* Formula:  @Transform(ComputedRunningZones; "x";
           @If(@IsMember(x; ComputedZoneAutobootTrue); @Nothing; x)) */
       public var ZonesNotRunningAutoboot:String = "";  /* Formula:  @Transform(ComputedNonRunningZones; "x";
           @If(@IsMember(x; ComputedZoneAutobootTrue); x; @Nothing)) */
```

Let @feather812002 know if you have other questions about how to use the formula properties.   I'll let you know if I get the example project working.

**Comment by piotrzarzycki21** [comment]
> @piotrzarzycki21, to clarify, are you seeing errors for the formulas on the Domino Visual Editor side? I see that the code snippet is invalid for Royale because of the unescaped double quotes.

I see errors only for Royale.

**Comment by piotrzarzycki21** [comment]
> Instead, the formula should be added as a comment, and the default value should be "". We can make improvements later to parse simple formulas (i.e literals).

I have moved formula value to comments. Let's discuss changes on today's call - build should be available by than with all of that.

**Comment by piotrzarzycki21** [comment]
`FormulaDefaultValue: ` - Add that in comment for formula

**Comment by piotrzarzycki21** [comment]
Stop populating default value for list of forms

**Comment by piotrzarzycki21** [comment]
Both changes has been apply. @JoelProminic I'm waiting for project which was failing for you during the meeting.

**Comment by JoelProminic** [comment]
@piotrzarzycki21, sorry for the delay.  @feather812002 is still working on debugging this issue - it is currently failing on one Base64-encoded defaultvalue attribute.   I removed that value for now, and the project opens in Moonshine. 

[xmlPrj_TestDefaultValue_incomplete.zip](https://github.com/Moonshine-IDE/Moonshine-IDE/files/9312382/xmlPrj_TestDefaultValue_incomplete.zip)

I'll put this at a more permanent location once I have the database fully working.

**Comment by piotrzarzycki21** [comment]
> @piotrzarzycki21, sorry for the delay. @feather812002 is still working on debugging this issue - it is currently failing on one Base64-encoded defaultvalue attribute. I removed that value for now, and the project opens in Moonshine.
> 
> [xmlPrj_TestDefaultValue_incomplete.zip](https://github.com/Moonshine-IDE/Moonshine-IDE/files/9312382/xmlPrj_TestDefaultValue_incomplete.zip)
> 
> I'll put this at a more permanent location once I have the database fully working.

In visualeditor-src folder of this project I have found weird file ".TestDefaultValue.xml.swp" - It was failing on an attempt of converting that one. I have fixed issue and now I would take into account only xml files - other one will be ignored.

**Comment by JoelProminic** [comment]
I did a quick test with the above code and the logic from #1074, and I see that the default values are not populated now.  This is fine for the 3.3.3 release, so I am going to bump this to 3.3.4
